docs/architecture/how/receipt-congestion.md

```diff
@@ -0,0 +1,211 @@
+
+The edges between shards are not limited in this model. In reality, we are
+eventually limited by the receipt sizes and what we can send within a block time
+through the network links. But if we only look at that limit, we can send very
+many receipts with a lot of gas attached to them. Thus, the model considers it
+unlimited.
+
+Okay, we have the capacities of the network modeled. Now let's look how a
+receipt execution maps onto it.
+
+Let's say a receipt starts at shard 1 with 300 Tgas. While executing, it burns 100 Tgas and
+creates an outgoing receipts with 200 Tgas to another shard. We can represent this in the flow network with 
+100 Tgas to the sink of shard 1 and 200 Tgas to shard 2. 
+
+![graph](../../images/congestion/receipt_flow_example_0.svg)
+
+Note: The graph includes the execution of the next block with the 200 Tgas to th
+sink of shard 2. This should be interpreted as if we continue sending the exact
+same workload on all shards every block. Then we reach this steady state where
+we continue to have these gas assignments per edge.
+
+Now we can do som flow analysis. It is immediately obvious that the total
+outflow per is limited to N * 1000 Tgas but the incoming flow is unlimited.
+
+For a finite amount of time, we can accept more inflow than outflow, we just have to add buffers to store what we cannot execute, yet. But to stay within finite memory requirements, we need to fall back to a flow diagram where outflows are greater or equal to inflows within a finite time frame.
+
+Next, we look at a ideas one at the time before combining some of them into the
+cross-shard congestion design proposed in
+[NEP-539](https://github.com/near/NEPs/pull/539).
+
+## Idea 1: Compute the minimum max-flow and stay below that limit
+
+One approach to solve congestion would be to never allow more work into the
+system than we can execute.
+
+But this is not ideal Just consider this example where everybody tries to access
+a contract on the same shard.
+
+![graph](../../images/congestion/receipt_flow_example_1.svg)
+
+In this workload where everyone want to use the capacity of the same shard, the
+max-flow of the system is essentially the 1000 Tgas that shard 3 can execute. No
+matter how many additional shards we add, this 1000 Tgas does not increase.
+
+Consequently, if we want to limit inflow to be the same or lower than the
+outflow, we cannot accept more than `1000 Tgas / NUM_SHARDS` of new transactions
+per chunk.
+
+![graph](../../images/congestion/receipt_flow_example_1_1.svg)
+
+So, can we just put a constant limit on sources that's `1000 Tgas / NUM_SHARDS`? Not
+really, as this limit is hardly practical. It means we limit global throughput
+to that of a single shard. Then why would we do sharding in the first place?
+
+The sad thing is, there is no way around it in the most general case. A
+congestion control strategy that does apply this limit to this workload will
+always have infinitely sized queues.
```

Comment:
It's the other way around, a strategy that accepts more than this limit will have unbounded queues.

Reply:
Right^^ I think a "not" went missing somewhere, I'll add one to correct it.